scripts/local-universe.py

```diff
@@ -115,38 +171,121 @@ def enumerate_docker_images(package_path):
     return (name for _, name in dockers.items())
 
 
-def download_docker_image(name, use_sudo):
+@contextlib.contextmanager
+def run_docker_registry(volume_path):
+    print('Start docker registry.')
+    command = [ 'docker', 'run', '-d', '-p', '5000:5000', '--name',
+        'registry', '-v', '{}:/var/lib/registry'.format(volume_path),
+        'registry:2.4.0']
+
+    subprocess.check_call(command)
+
+    try:
+        yield
+    finally:
+        print('Stopping docker registry.')
+        command = [ 'docker', 'rm', '-f', 'registry']
+        subprocess.call(command)
+
+
+def download_docker_image(name):
     print('Pull docker images: {}'.format(name))
     command = ['docker', 'pull', name]
 
-    command = ['sudo'] + command if use_sudo else command
-    subprocess.call(command)
+    subprocess.check_call(command)
 
 
-def add_http_resource(zip_file, url, dir_path):
-    archive_path = (dir_path /
-                    pathlib.PurePosixPath(urllib.parse.urlparse(url).path).name)
-    print('Adding {} at {} to file zip.'.format(url, archive_path))
-    with urllib.request.urlopen(url) as response:
-        with tempfile.NamedTemporaryFile() as tempFile:
-            shutil.copyfileobj(response, tempFile)
-            tempFile.flush()
-            zip_file.write(tempFile.name, str(archive_path))
+def format_image_name(host, name):
+    # Probably has a hostname at the front, get rid of it.
+    if '.' in name.split(':')[0]:
+        return '{}/{}'.format(host, "/".join(name.split("/")[1:]))
 
+    return '{}/{}'.format(host, name)
```

Comment:
I think we can safely use `/` as separators for `[(host:port)]/user/name`. `/` are not valid characters for `user` and `name`.

> The image name which is made up of slash-separated name components, optionally prefixed by a registry hostname. The hostname must comply with standard DNS rules, but may not contain underscores. If a hostname is present, it may optionally be followed by a port number in the format `:8080`. If not present, the command uses Docker's public registry located at `registry-1.docker.io` by default. Name components may contain lowercase characters, digits and separators. A separator is defined as a period, one or two underscores, or one or more dashes. A name component may not start or end with a separator.